s3repo/model.py

```diff
@@ -378,9 +388,103 @@ def delete_package(self, package):
+        directory from S3.
+        """
+
+        paginator = self.s3_client.get_paginator('list_objects')
+        list_parameters = {'Bucket': self.bucket.name,
+                           'Prefix': path}
+
+        files = {}
+        folders = {}
+        try:
+            for result in paginator.paginate(**list_parameters):
+                if result.get('Contents') is not None:
+                    for fileobj in result.get('Contents'):
+                        # Select the names of all files and
+                        # folders inside the current one.
+                        directory_path_len = 0
+                        if len(path) != 0:
+                            directory_path_len = len(path)
+                        tail = fileobj.get('Key')[directory_path_len:]
+                        item_name = tail.split('/')[0]
+                        last_modified = fileobj.get('LastModified')
+
+                        # list_objects() paginator returns the
+                        # current directory, as well as files with
+                        # the same name (prefix), messing with the
+                        # current directory, because they are S3
+                        # objects too. Ignore these objects.
+                        if item_name == '':
+                            continue
+                        # If tail is longer than item_name (i.e.
+                        # have /... afetr item_name), it means
+                        # that we deal with directory.
+                        if tail != item_name:
+                            in_list = item_name in folders
+                            if (not in_list or in_list
+                                    and folders[item_name] < last_modified):
+                                folders[item_name] = last_modified
+                            folders[item_name] = last_modified
+                        else:
+                            size = fileobj.get('Size')
+
+                            # Make size human readable.
+                            human_readadble = ''
+                            for unit in ['','Ki','Mi','Gi']:
+                                if abs(size) < 1024.0:
+                                    human_readadble = "%3.1f %sB" % (size, unit)
+                                    break
+                                size /= 1024.0
+
+                            files[item_name] = {'last_modified': last_modified,
+                                                'size': human_readadble}
+                else:
+                    raise RuntimeError("No such folder.")
+        except self.s3_client.exceptions.NoSuchBucket:
+            raise RuntimeError("No such bucket.")
+
+        return (files, folders)
+
     def get_file(self, path):
-        """Download a file from S3."""
-        NotImplementedError("get_file hasn't been implemented yet.")
+        """Get a file from S3."""
+
+        self.sync_lock.acquire()
```

Comment:
This mutex is used to protect "unsync_repos".
Why do you need mutex at all?

Reply:
I think files while downloading may be a competing stream.